File: frontend/src/services/portfolio.ts
import { api } from "./api";

export interface PositionOut {
  asset_id: number;
  symbol: string;
  name: string;
  asset_type: string;
  currency: string;
  exchange: string;
  quantity: number;
  pmc_eur: number;
  total_invested_eur: number;
  realized_pnl_eur: number;
  current_price: number | null;
  current_price_eur: number | null;
  current_value_eur: number | null;
  unrealized_pnl_eur: number | null;
  unrealized_pnl_pct: number | null;
  change_pct: number | null;
  period_pnl_eur: number | null;
  period_pnl_pct: number | null;
}

export interface PortfolioSummaryOut {
  total_value_eur: number;
  total_invested_eur: number;
  total_pnl_eur: number;
  total_pnl_pct: number;
  realized_pnl_eur: number;
  unrealized_pnl_eur: number;
  daily_change_eur: number;
  positions_count: number;
}

export interface PerformancePoint {
  date: string;
  value_eur: number;
  invested_eur: number;
  pnl_eur: number;
  twrr_pct: number;
}

export interface PerformanceOut {
  period: string;
  twrr_pct: number;
  series: PerformancePoint[];
}

export interface AllocationItem {
  label: string;
  value_eur: number;
  pct: number;
  count: number;
}

export interface AllocationOut {
  by_type: AllocationItem[];
  by_currency: AllocationItem[];
  by_account: AllocationItem[];
  by_sector: AllocationItem[];
  by_continent: AllocationItem[];
  total_value_eur: number;
}

export interface ETFHoldingItem {
  symbol: string;
  name: string;
  weight: number;
}

export interface CountryItem {
  code: string;
  name: string;
  weight: number;
}

export interface ETFHoldingOut {
  asset_id: number;
  symbol: string;
  name: string;
  value_eur: number | null;
  holdings: ETFHoldingItem[];
  is_override: boolean;
  countries_override: CountryItem[] | null;
}

export interface DashboardOut {
  summary: PortfolioSummaryOut;
  positions: PositionOut[];
  allocation: AllocationOut;
}

export interface DividendOut {
  id: number;
  date: string;
  asset_id: number;
  symbol: string;
  name: string;
  type: string;
  amount_eur: number;
  account_name: string;
  account_id: number;
}

export interface HoldingPricePoint {
  date: string;
  price: number;
}

export interface HoldingActivityOut {
  id: number;
  type: string;
  date: string;
  quantity: number;
  price: number;
  price_currency: string;
  total_eur: number;
  fee: number;
  account_name: string;
  account_id: number;
}

export interface HoldingAccountOut {
  account_id: number;
  account_name: string;
  quantity: number;
  value_eur: number | null;
  pct: number | null;
}

export interface SectorItem {
  name: string;
  weight: number;
}

export interface HoldingDetailOut {
  asset_id: number;
  symbol: string;
  name: string;
  asset_type: string;
  currency: string;
  exchange: string;
  isin: string | null;
  quantity: number;
  pmc_eur: number;
  total_invested_eur: number;
  realized_pnl_eur: number;
  current_price: number | null;
  current_price_eur: number | null;
  current_value_eur: number | null;
  unrealized_pnl_eur: number | null;
  unrealized_pnl_pct: number | null;
  change_pct: number | null;
  min_price: number | null;
  max_price: number | null;
  total_fees: number;
  activities_count: number;
  first_buy_date: string | null;
  price_history: HoldingPricePoint[];
  activities: HoldingActivityOut[];
  accounts: HoldingAccountOut[];
  sectors: SectorItem[] | null;
  countries: CountryItem[] | null;
}

export interface RiskMetricsOut {
  period: string;
  trading_days: number;
  annualized_volatility_pct: number;
  max_drawdown_pct: number;
  max_drawdown_start: string | null;
  max_drawdown_end: string | null;
  sharpe_ratio: number | null;
  sortino_ratio: number | null;
  calmar_ratio: number | null;
  twrr_annualized_pct: number;
  best_day_pct: number;
  worst_day_pct: number;
  positive_days_pct: number;
}

export const portfolioService = {
  async getDashboard(period = "ytd"): Promise<DashboardOut> {
    const { data } = await api.get<DashboardOut>("/portfolio/dashboard", { params: { period } });
    return data;
  },

  async getSummary(): Promise<PortfolioSummaryOut> {
    const { data } = await api.get<PortfolioSummaryOut>("/portfolio/summary");
    return data;
  },

  async getPositions(): Promise<PositionOut[]> {
    const { data } = await api.get<PositionOut[]>("/portfolio/positions");
    return data;
  },

  async getPerformance(period: string, accountId?: number): Promise<PerformanceOut> {
    const { data } = await api.get<PerformanceOut>("/portfolio/performance", {
      params: { period, ...(accountId != null ? { account_id: accountId } : {}) },
    });
    return data;
  },

  async getAllocation(): Promise<AllocationOut> {
    const { data } = await api.get<AllocationOut>("/portfolio/allocation");
    return data;
  },

  async getDividends(): Promise<DividendOut[]> {
    const { data } = await api.get<DividendOut[]>("/portfolio/dividends");
    return data;
  },

  async getHoldingDetail(assetId: number): Promise<HoldingDetailOut> {
    const { data } = await api.get<HoldingDetailOut>(`/portfolio/holding/${assetId}`);
    return data;
  },

  async getEtfHoldings(): Promise<ETFHoldingOut[]> {
    const { data } = await api.get<ETFHoldingOut[]>("/portfolio/etf-holdings");
    return data;
  },

  async getRisk(period: string, accountId?: number): Promise<RiskMetricsOut> {
    const { data } = await api.get<RiskMetricsOut>("/portfolio/risk", {
      params: { period, ...(accountId != null ? { account_id: accountId } : {}) },
    });
    return data;
  },

  async getXirr(): Promise<{ xirr_pct: number | null }> {
    const { data } = await api.get<{ xirr_pct: number | null }>("/portfolio/xirr");
    return data;
  },

  async getBenchmark(index: string, period: string): Promise<{ index: string; ticker: string; period: string; series: { date: string; value: number }[] }> {
    const { data } = await api.get("/portfolio/benchmark", { params: { index, period } });
    return data;
  },

  async getCorrelation(period: string): Promise<{ labels: string[]; matrix: number[][]; period: string }> {
    const { data } = await api.get("/portfolio/correlation", { params: { period } });
    return data;
  },

  async getCountryAllocation(): Promise<{
    countries: { code: string; name: string; value_eur: number; pct: number; market_type: "developed" | "emerging" | "other" }[];
    totals: { developed_pct: number; emerging_pct: number; other_pct: number; no_data_pct: number };
  }> {
    const { data } = await api.get("/portfolio/country-allocation");
    return data;
  },
};

File: backend/app/schemas/portfolio.py
import datetime

from pydantic import BaseModel


class PositionOut(BaseModel):
    asset_id: int
    symbol: str
    name: str
    asset_type: str
    currency: str
    exchange: str
    quantity: float
    pmc_eur: float
    total_invested_eur: float
    realized_pnl_eur: float
    current_price: float | None
    current_price_eur: float | None
    current_value_eur: float | None
    unrealized_pnl_eur: float | None
    unrealized_pnl_pct: float | None
    change_pct: float | None
    period_pnl_eur: float | None = None
    period_pnl_pct: float | None = None


class PortfolioSummaryOut(BaseModel):
    total_value_eur: float
    total_invested_eur: float
    total_pnl_eur: float
    total_pnl_pct: float
    realized_pnl_eur: float
    unrealized_pnl_eur: float
    daily_change_eur: float
    positions_count: int


class PerformancePoint(BaseModel):
    date: datetime.date
    value_eur: float
    invested_eur: float
    pnl_eur: float
    twrr_pct: float


class PerformanceOut(BaseModel):
    period: str
    twrr_pct: float
    series: list[PerformancePoint]


class AllocationItem(BaseModel):
    label: str
    value_eur: float
    pct: float
    count: int


class AllocationOut(BaseModel):
    by_type: list[AllocationItem]
    by_currency: list[AllocationItem]
    by_account: list[AllocationItem]
    by_sector: list[AllocationItem] = []
    by_continent: list[AllocationItem] = []
    total_value_eur: float


class ETFHoldingItem(BaseModel):
    symbol: str
    name: str
    weight: float


class CountryItem(BaseModel):
    code: str
    name: str
    weight: float


class ETFHoldingOut(BaseModel):
    asset_id: int
    symbol: str
    name: str
    value_eur: float | None
    holdings: list[ETFHoldingItem]
    is_override: bool
    countries_override: list[CountryItem] | None = None


class DashboardOut(BaseModel):
    summary: PortfolioSummaryOut
    positions: list[PositionOut]
    allocation: AllocationOut


class DividendOut(BaseModel):
    id: int
    date: datetime.date
    asset_id: int
    symbol: str
    name: str
    type: str
    amount_eur: float
    account_name: str
    account_id: int


class RiskMetricsOut(BaseModel):
    period: str
    trading_days: int
    annualized_volatility_pct: float
    max_drawdown_pct: float
    max_drawdown_start: datetime.date | None
    max_drawdown_end: datetime.date | None
    sharpe_ratio: float | None
    sortino_ratio: float | None
    calmar_ratio: float | None
    twrr_annualized_pct: float
    best_day_pct: float
    worst_day_pct: float
    positive_days_pct: float


class HoldingPricePoint(BaseModel):
    date: datetime.date
    price: float


class HoldingActivityOut(BaseModel):
    id: int
    type: str
    date: datetime.date
    quantity: float
    price: float
    price_currency: str
    total_eur: float
    fee: float
    account_name: str
    account_id: int


class HoldingAccountOut(BaseModel):
    account_id: int
    account_name: str
    quantity: float
    value_eur: float | None
    pct: float | None


class SectorItem(BaseModel):
    name: str
    weight: float


class HoldingDetailOut(BaseModel):
    asset_id: int
    symbol: str
    name: str
    asset_type: str
    currency: str
    exchange: str
    isin: str | None
    quantity: float
    pmc_eur: float
    total_invested_eur: float
    realized_pnl_eur: float
    current_price: float | None
    current_price_eur: float | None
    current_value_eur: float | None
    unrealized_pnl_eur: float | None
    unrealized_pnl_pct: float | None
    change_pct: float | None
    min_price: float | None
    max_price: float | None
    total_fees: float
    activities_count: int
    first_buy_date: datetime.date | None
    price_history: list[HoldingPricePoint]
    activities: list[HoldingActivityOut]
    accounts: list[HoldingAccountOut]
    sectors: list[SectorItem] | None = None
    countries: list[CountryItem] | None = None
